File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.py
from sys import stdin as s
from collections import deque


#s=open("input.txt","rt")


def bfs(graph,start,visited,k):
    
    queue = deque()
    queue.append(start) #큐에 시작 값을 넣어줌
    while queue:
        v= queue.popleft()
        visited[start]=True #visited에 방문 값을 true로 바꿔줌
        for i in graph[v]: 
            if not visited[i]: #만약 노드에 방문하지 않았다면
                queue.append(i) #큐에 노드를 추가하고
                visited[i]=True # 방문 true로 해준다음 
                go_list[i] = go_list[v]+1 #간선의 가중치 값을 올려줌

           

n,m,k,x = map(int,s.readline().split()) #n : 도시, m은 간선, k:최단 거리, x : 출발도시
list1 =[]
graph=[[] for _ in range(n+1)]
visited = [False]*(n+1) #방문 여부를 입력해줌
go_list = [0 for _ in range(n+1)] # 간선의 가중치를 저장해줌 
#도달할 수 있는 도시 입력받고 graph 배열에 1,2,3,4 차례로 넣어주기
for i in range(m):
    a,b=map(int,s.readline().split())
    graph[a].append(b)
    graph[a].sort()
bfs(graph,x,visited,k)
    


                
count=0
for i in range(1,n+1):
    if go_list[i]==k:
        print(i)
        count+=1
if count ==0:
    print(-1)    

File: 백준/Silver/18352. 특정 거리의 도시 찾기/특정 거리의 도시 찾기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static ArrayList<Integer> A[];
    static int N;
    static int M;
    static int k;
    static int x;
    static boolean visited[];
    static int go_list[];
    static List<Integer> answer;
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        x = Integer.parseInt(st.nextToken());
        answer = new ArrayList<>();
        visited = new boolean[N+1];

        A = new ArrayList[N+1];
        for(int i=1; i<=N; i++){
            A[i]=new ArrayList<Integer>();
        }

        for(int i=0;i<M; i++){
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            A[a].add(b);
        }

        go_list = new int[N+1];
        for(int i=0; i<=N;i++){
            go_list[i]=-1;
        }

       
        BFS(x);
        for(int i=0; i<=N; i++){
            if(go_list[i]==k){
                answer.add(i);   
            }
        }
        
        if(answer.isEmpty()){
            System.out.print("-1");
        }
        else{
            Collections.sort(answer); //collections과 array의 차이를 잘 모르겠음.
            for(int a: answer){
                System.out.println(a);
            }   
        }
    }
    private static void BFS(int node){
        Queue<Integer> queue = new LinkedList<Integer>();
        queue.add(node);
        go_list[node]++;

        while(!queue.isEmpty()){
            int now_node = queue.poll();
            for(int i: A[now_node]){
                if(go_list[i]==-1){
                    go_list[i]=go_list[now_node]+1;
                    queue.add(i);
                }
            }
        }
    }
}
